lib/modules/datasource/deno/index.ts

```diff
@@ -0,0 +1,109 @@
+import is from '@sindresorhus/is';
+import pMap from 'p-map';
+import { logger } from '../../../logger';
+import { cache } from '../../../util/cache/package/decorator';
+import { regEx } from '../../../util/regex';
+import { joinUrlParts } from '../../../util/url';
+import * as semanticVersioning from '../../versioning/semver';
+import { Datasource } from '../datasource';
+import type { Release } from '../index';
+import type { GetReleasesConfig, ReleaseResult } from '../types';
+import type {
+  DenoAPIModuleResponse,
+  DenoAPIModuleVersionResponse,
+} from './types';
+import { createSourceURL, tagsToRecord } from './utils';
+
+export class DenoDatasource extends Datasource {
+  static readonly id = 'deno';
+
+  override readonly customRegistrySupport = true;
+
+  override readonly registryStrategy = 'first';
+
+  override readonly defaultVersioning = semanticVersioning.id;
+
+  override readonly defaultRegistryUrls = ['https://apiland.deno.dev'];
+
+  constructor() {
+    super(DenoDatasource.id);
+  }
+
+  @cache({
+    namespace: `datasource-${DenoDatasource.id}`,
+    key: ({ packageName, registryUrl }: GetReleasesConfig) =>
+      // TODO: types (#7154)
+      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+      `${registryUrl}:${packageName}`,
+  })
+  async getReleases({
+    packageName,
+    registryUrl,
+  }: GetReleasesConfig): Promise<ReleaseResult | null> {
+    const massagedRegistryUrl = registryUrl ?? this.defaultRegistryUrls[0];
+
+    const extractResult = regEx(
+      '^(https:\\/\\/deno.land\\/)(?<rawPackageName>[^@\\s]+)'
+    ).exec(packageName);
+    if (is.nullOrUndefined(extractResult)) {
+      logger.debug(
+        `Could not match package regex against packageName: "${packageName}"`
+      );
+      return null;
+    }
+
+    const rawPackageName = extractResult.groups?.rawPackageName;
+    // istanbul ignore if
+    if (is.nullOrUndefined(rawPackageName)) {
+      logger.debug(
+        `Could not extract rawPackageName from package regex: "${packageName}"`
+      );
+      return null;
+    }
+
+    // remove third-party prefix if defined. The only internal library is `std` and is available under the same API
+    const massagedPackageName = rawPackageName.replace('x/', '');
+
+    // https://apiland.deno.dev/v2/modules/postgres
+    const moduleAPIURL = joinUrlParts(
+      massagedRegistryUrl,
+      'v2/modules',
+      massagedPackageName
+    );
+    const { versions, tags } = (
+      await this.http.getJson<DenoAPIModuleResponse>(moduleAPIURL)
+    ).body;
+
+    // get details for the versions
+    const releases = await pMap(
```

Comment:
This could be a *lot* of lookups. Are the results for each version considered immutable? If so then we would only need to look up versions not in the cache already

Reply:
Yes, they are immutable. The caching happens here: https://github.com/renovatebot/renovate/pull/18878/files#diff-ad918d06da5c9ca341613d30167bda1b44146a2ee132cd9166b340e315edbbb0R98

Reply:
Sorry, missed that. However this will still be very "heavy" in terms of cache requests (one cache request per version). Ideally we would have a single cache to storing all versions we know about.

Reply:
I see. I will put in another caching layer between the current ones.

Reply:
Great, please ping me again once done

Reply:
I have added a caching layer for the version endpoint.

Reply:
I'm starting to reach the limits of my code review capabilities, but I'm not sure if this is caching as I'd hoped.

Let's say on one run you query a certain package and get back 10 versions. In this case it would have taken us 11 API calls? i.e. 1 to get the list of versions and then an additional call per version to get the details? How many calls to cache?

Now let's say you run again in an hour, query Deno land again, and get back the same 10 versions. Any additional calls to Deno land? How many calls to cache?

Reply:
The first cache is based on the requested `packageName` as well as `registryUrl`. 
The second one is based on the module API url and wraps all API calls.     
Should both be invalidated, an API call to the API server is executed to get the newest version list.     
With this list of versions, the details for each version are requested.  Most if not all versions are then retrieved from the cache as these are cached for a week. The only details from the external API requested in this time frame are the ones which are newly added.

Reply:
Assume that the initial cache is invalidated due to normal reasons (e.g. it's an hour or so old). Assume there's 100 releases for this package, none of which have changed. We'll do an additional 100 requests to the cache (1 per release)? I would instead prefer to cache release details _in one entry_, so we have only one additional request to the cache. Does that make sense?

Reply:
I have implemented now a custom cache logic to prevent this. 

Basically all version details are stored in a single cache and are updated if necessary. Therefore there are at max 4 calls to the caching service ( 3 GET & 1 POST )

IMO it would be worth to extend central caching with a similar functionality later on.